File: pset6/crack.py
import sys
import crypt
import string
import time
from hmac import compare_digest as compare_hash

#character array and its size in a variable
chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
charsNum = len(chars)

def main():

    #check system arguments
    if(len(sys.argv) != 2):
        print("Usage: python crack.py hash")
        exit(1)
    
    #store hash
    inputHash = sys.argv[1]
    
    #max number of length
    maxLength = 4
    
    #loop integer
    i = 1
    
    begin = time.process_time()
    #check every word from length 1 - maxLength
    for i in range(maxLength + 1):
        wordGenerator(i, inputHash)
        
    end = time.process_time()
    
    print("elapsed time {}".format(end - begin))
    
#generates all words of a given length
def wordGenerator(wordLength, hashWord):
    
    #declare chars and charsNum as global variable
    global chars, charsNum
    
    #loop integer, entry array
    entry = [0] * wordLength
    
    #word array
    word = [None] * wordLength
    
    #loop to generate the words
    while True:
        
        #sets word elements to the alphabet of position entry[i]
        for i in range(wordLength):
            word[i] = chars[int(entry[i])]
        
        #check the password
        if checkPass(''.join(word), hashWord):
            print("Correct password: {}".format(''.join(word)))
            return
            
        
        j = 0
        #update entry elements if needed
        while(j < wordLength):
            #increment entry
            entry[j] += 1
            
            #if entry[i] hasn't reached 'z' then exit the loop
            if charsNum is not entry[j]:
                break
            
            #else set it to 0
            entry[j] = 0
            
            #increment loop variable
            j += 1
            
        #breaks out of loop when all words are generated
        if(j >= wordLength):
            break
    
#check if pass is true    
def checkPass(generatedPass, userHash):
    
    #generate hash from the given password
    generatedHash = crypt.crypt(generatedPass, "50")
    
    #return the boolean value
    return compare_hash(generatedHash, userHash)

    
if __name__ == "__main__":
    main()

File: pset2/crack/crack.c
#define _XOPEN_SOURCE
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>

void wordGenerator(int wordLength, const char* inputHash);
bool checkPass(const char* key, const char* hash);


int len = 4;
int count = 0;
char letters[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
int nbletters = sizeof(letters)-1;

int main(int argc, char* argv[]) {
    
    //exits when command line arguments aren't correct
    if(argc != 2){
        fprintf(stderr, "Usage: ./crack hash\n");
        return 1;
    }
    
    //stores given hash    
    char* userHash = argv[1];
 
    //tries out all words of length from 1 - len
    clock_t begin = clock();
    for(int i = 1; i <= len; i++){
        wordGenerator(i, userHash);
    }
    clock_t end = clock();
    
    double timeSpent = (double) (end - begin) / CLOCKS_PER_SEC;
    
    printf("Time spent:  %f\n", timeSpent);
}

//generates all the words of a given length
void wordGenerator(int wordLength, const char* inputHash){
    
    //loop integer, array to store letter pos
    int i, entry[wordLength];
    
    //array to store word
    char pass[wordLength + 1];
    
    //initialiazes entry array
    for(i = 0 ; i < wordLength ; i++) 
    	entry[i] = 0;
    
    //loop to generate words 	
    do {
        //sets pass[i] to the alphabet of position entry[i]
        //for the given length
        for(i = 0 ; i < wordLength ; i++) 
        	pass[i] = letters[entry[i]];
        	
        //stops string from being combined with elements in memory
        pass[wordLength] = '\0';
        
        printf("%s\n", pass);
        
        //exits when password is given 
        if(checkPass(pass, inputHash)){
            fprintf(stdout, "Correct password: %s\n", pass);
            return;
        }
        
        //sets entry elemnts to 0 when they've reached the limit
        for(i = 0 ; i < wordLength && ++entry[i] == nbletters; i++)
        	entry[i] = 0;
        
    } while(i < wordLength);
    
    
}

bool checkPass(const char* password, const char* hash){
    if(strcmp(crypt(password, "50"), hash) == 0){
            return true;
    }
    return false;
}